symengine/polys/basic_to_poly.cpp

```diff
@@ -0,0 +1,185 @@
+    void add_to_gen_set(const RCP<const Basic> &base,
+                        const RCP<const Number> &exp)
+    {
+        auto it = gen_set.find(base);
+        if (it == gen_set.end()) {
+            gen_set[base] = exp;
+            return;
+        }
+
+        if (is_a<const Rational>(*exp) and is_a<const Rational>(*it->second)) {
+            gen_set[base] = rcp_static_cast<const Rational>(div(
+                one,
+                lcm(*rcp_static_cast<const Rational>(exp)->get_den(),
+                    *rcp_static_cast<const Rational>(it->second)->get_den())));
+        } else if (is_a<const Rational>(*exp)) {
+            gen_set[base] = rcp_static_cast<const Rational>(
+                div(one, rcp_static_cast<const Rational>(exp)->get_den()));
+        }
+    }
+
+    void bvisit(const Pow &x)
+    {
+        if (is_a<const Integer>(*x.get_exp())) {
+            if (rcp_static_cast<const Integer>(x.get_exp())->is_positive()) {
+                x.get_base()->accept(*this);
+            } else {
+                add_to_gen_set(pow(x.get_base(), minus_one), one);
+            }
+
+        } else if (is_a<const Rational>(*x.get_exp())) {
+
+            RCP<const Basic> mulx = one;
+            if (rcp_static_cast<const Rational>(x.get_exp())->is_negative())
+                mulx = minus_one;
+            RCP<const Rational> den = rcp_static_cast<const Rational>(div(
+                one, rcp_static_cast<const Rational>(x.get_exp())->get_den()));
+            add_to_gen_set(pow(x.get_base(), mulx), den);
+
+        } else {
+            umap_basic_num pow_pairs
+                = _find_gens_poly_pow(x.get_exp(), x.get_base());
+            for (auto it : pow_pairs)
+                add_to_gen_set(pow(x.get_base(), it.first), it.second);
+        }
+    }
+
+    void bvisit(const Add &x)
+    {
+        for (auto it : x.dict_)
+            it.first->accept(*this);
+    }
+
+    void bvisit(const Mul &x)
+    {
+        for (auto it : x.dict_)
+            it.first->accept(*this);
+    }
+
+    void bvisit(const Number &x)
+    {
+        // intentionally blank
+    }
+
+    void bvisit(const Basic &x)
+    {
+        add_to_gen_set(x.rcp_from_this(), one);
+    }
+};
+
+class PolyGeneratorVisitorPow : public BaseVisitor<PolyGeneratorVisitorPow>
+{
+private:
+    // the generators are mul(it.first, it.second) not Pow
+    // the_base is the base of the Pow (whose exp we are currently dealing)
+    // numbers must always be positive and of the form (1/d, d belongs to N)
+    umap_basic_num gen_set;
+    RCP<const Basic> the_base;
+
+public:
+    umap_basic_num apply(const Basic &b, const RCP<const Basic> &base)
+    {
+        the_base = base;
+        b.accept(*this);
+        return gen_set;
+    }
+
+    void bvisit(const Pow &x)
+    {
+        umap_basic_num pow_pairs
+            = _find_gens_poly_pow(x.get_exp(), x.get_base());
+        for (auto it : pow_pairs)
+            gen_set[pow(x.get_base(), it.first)] = it.second;
```

Comment:
Take `2**(3*x + x**2 + (y/2))`
`find_gen(2**(3*x + x**2 + (y/2))`
`= 2** find_gen_pow(3*x + x**2 + (y/2))`
`= 2** (x, x**2, y/2)`
`= (2**x, 2**x**2, 2**(y/2))`

So, in  a way `find_gen_pow` does with coefficients what `find_gen` does with exponents. As in `find_gen` the exponents are stored for each generator, and the exponents is always of the type `1/d` where `d` belongs to natural numbers, similarly for `find_gen_pow` the coefficients are always stored in `1/d` form.

Thus like in `find_gen` we can't have `x` and `x**1/2` both, because one is square of other, we can't have `x` and `x/2` both in `find_gen_pow` as in exponent terms, one is square of the other.